src/imageio/format/jxl.c

```diff
@@ -0,0 +1,637 @@
+
+  // Set encoder basic info
+  JxlBasicInfo basic_info;
+  JxlEncoderInitBasicInfo(&basic_info);
+  basic_info.xsize = width;
+  basic_info.ysize = height;
+  switch(params->bpp)
+  {
+    case 0:
+      basic_info.bits_per_sample = 8;
+      basic_info.exponent_bits_per_sample = 0;
+      break;
+    case 1:
+      basic_info.bits_per_sample = 10;
+      basic_info.exponent_bits_per_sample = 0;
+      break;
+    case 2:
+      basic_info.bits_per_sample = 12;
+      basic_info.exponent_bits_per_sample = 0;
+      break;
+    case 3:
+      basic_info.bits_per_sample = 16;
+      basic_info.exponent_bits_per_sample = 0;
+      break;
+    case 4:
+      basic_info.bits_per_sample = 16;
+      basic_info.exponent_bits_per_sample = 5;
+      break;
+    default:
+      basic_info.bits_per_sample = 32;
+      basic_info.exponent_bits_per_sample = 8;
+      break;
+  }
+  // Lossless only makes sense for integer modes
+  if(basic_info.exponent_bits_per_sample == 0 && params->quality == 100)
+  {
+    // Must preserve original profile for lossless mode
+    basic_info.uses_original_profile = JXL_TRUE;
+    LIBJXL_ASSERT(JxlEncoderSetFrameDistance(frame_settings, 0.0f));
+    LIBJXL_ASSERT(JxlEncoderSetFrameLossless(frame_settings, JXL_TRUE));
+  }
+  else
+  {
+    basic_info.uses_original_profile = params->original == FALSE ? JXL_FALSE : JXL_TRUE;
+    const float distance = params->quality >= 30 ? 0.1f + (100 - params->quality) * 0.09f
+                                           : 6.24f + powf(2.5f, (30 - params->quality) / 5.0f) / 6.25f;
+    LIBJXL_ASSERT(JxlEncoderSetFrameDistance(frame_settings, MIN(distance, 25.0f)));
+  }
+
+  LIBJXL_ASSERT(JxlEncoderFrameSettingsSetOption(frame_settings, JXL_ENC_FRAME_SETTING_EFFORT, params->effort));
+
+  LIBJXL_ASSERT(
+      JxlEncoderFrameSettingsSetOption(frame_settings, JXL_ENC_FRAME_SETTING_DECODING_SPEED, params->tier));
+
+  // Codestream level should be chosen automatically given the settings
+  LIBJXL_ASSERT(JxlEncoderSetBasicInfo(encoder, &basic_info));
+
+  // Determine and set the encoder color space
+  const dt_colorspaces_color_profile_t *output_profile
+      = dt_colorspaces_get_output_profile(imgid, over_type, over_filename);
+  const cmsHPROFILE out_profile = output_profile->profile;
+  // Previous call will give us a more accurate color profile type
+  // (not what the user requested in the export menu but what the image is actually using)
+  over_type = output_profile->type;
+
+  // If possible we want libjxl to save the color encoding in its own format, rather
+  // than as an ICC binary blob which is possible.
+  // If we are unable to find the required color encoding data for libjxl we will
+  // just fallback to providing an ICC blob (and hope we can at least do that!).
+  bool write_color_natively = true;
+
+  JxlColorEncoding color_encoding;
+  color_encoding.color_space = JXL_COLOR_SPACE_RGB;
```

Comment:
I would add here:
`color_encoding.white_point = JXL_WHITE_POINT_D65;`

as this is common to all types and removed the setting in the `switch()` below.

Reply:
Done